### PR TITLE
Issue 5658: Move spine builders to separate jar file

### LINKED ISSUE #5658: Move SpineSceneBuilder et al to a separate jar file
- [ ] Move the Spine scene building to a different source location.
- [ ] Use `fmt-spine.jar` in bob.jar
- [ ] Use `fmt-spine.jar` in Editor

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java [hunk 1]
```diff
@@ -457,8 +471,22 @@ else if (severity == MultipleCompileException.Info.SEVERITY_WARNING)
...
+        } catch(ClassNotFoundException e) {
+            return null;
+        }
+    }
+
     private static void setupProject(Project project, boolean resolveLibraries, String sourceDirectory) throws IOException, LibraryException, CompileExceptionError {
-        ClassLoaderScanner scanner = new ClassLoaderScanner();
+
+        // Find the jar file in the built-in resources
+        String jar = Bob.getJarFile("fmt-spine.jar");
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java:485]: Finding these from the native extensions later on

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ClassLoaderScanner.java [hunk 2]
```diff
@@ -62,28 +70,55 @@ protected URL resolveURL(URL url) throws IOException {
...
                 }
             }
         } catch (IOException e) {
             throw new RuntimeException(e);
         }
+    }
+
+    @Override
+    public Set<String> scan(String pkg) {
+        System.out.printf("ClassLoaderScanner: scan() packageName: %s\n", pkg);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ClassLoaderScanner.java]: Should be removed

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ProtoBuilder.java [hunk 3]
```diff
@@ -25,8 +27,37 @@
...
     public ProtoBuilder() {
         protoParams = getClass().getAnnotation(ProtoParams.class);
+
+        BuilderParams builderParams = getClass().getAnnotation(BuilderParams.class);
+        extToMessageClass.put(builderParams.outExt(), protoParams.messageClass());
+    }
+
+    static public void addMessageClass(String ext, Class<? extends GeneratedMessage> klass) {
+        extToMessageClass.put(ext, klass);
+        System.out.printf("addMessageClass ext: '%s'  class: '%s'\n", ext, klass.getName());
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ProtoBuilder.java]: Remove

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ProtoBuilder.java [hunk 4]
```diff
@@ -25,8 +27,37 @@
 
     private ProtoParams protoParams;
 
+    private static Map<String, Class<? extends GeneratedMessage>> extToMessageClass = new HashMap<String, Class<? extends GeneratedMessage>>();
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ProtoBuilder.java:30]: This mechanic is to allow for a dynamic registration of protobuf message builders, as opposed to the hard coded list we had.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/GuiBuilder.java [hunk 5]
```diff
@@ -573,55 +576,80 @@ else if(node.getType() == Type.TYPE_SPINE) {
...
-                                parentId = spineNodeId;
-                            }
-                            boneNodeBuilder.setParent(parentId);
-
-                            NodeDesc boneNode = boneNodeBuilder.build();
-                            newScene.get("").add(boneNode);
-                            for(LayoutDesc layout : sceneBuilder.getLayoutsList()) {
-                                newScene.get(layout.getName()).add(boneNode);
-                            }
+                        Method getSpineJson = spineSceneBuilder.getClass().getDeclaredMethod("getSpineJson");
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/GuiBuilder.java:587]: Part of the api for the new Spine plugin

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/GuiBuilder.java [hunk 6]
```diff
@@ -573,55 +576,80 @@ else if(node.getType() == Type.TYPE_SPINE) {
...
+                        throw new CompileExceptionError(builder.project.getResource(input), 0, "Failed to find SpineSceneUtil class!");
+                    }
+                    Class<?> loadExceptionClass = Bob.getClass("com.dynamo.bob.pipeline.SpineSceneUtil$LoadException");
+                    if (loadExceptionClass == null) {
+                        throw new CompileExceptionError(builder.project.getResource(input), 0, "Failed to find SpineSceneUtil.LoadException class!");
+                    }
+
+                    List<RigUtil.Bone> bones;
+                    try {
+                        Method getBones = spineSceneUtilClass.getDeclaredMethod("getBones", InputStream.class);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/GuiBuilder.java:609]: Another part of the api, to get the bones from the scene.

File: editor/src/clj/editor/spine.clj [hunk 7]
```diff
@@ -39,17 +39,30 @@
...
+  (let [loader (clojure.lang.DynamicClassLoader. (.getContextClassLoader (Thread/currentThread)))
+        url (URL. (format "file://%s/shared/java/fmt-spine.jar" (system/defold-unpack-path)))]
+    (.addURL loader url)
+    loader))
+
+(defn dynamically-load-class!
+  [class-loader class-name]
+  (Class/forName class-name true class-loader))
+
+
```
Comment (JCash) [inline editor/src/clj/editor/spine.clj:65]: As mentioned, we need to centralize this later to.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java [hunk 8]
```diff
@@ -223,12 +223,24 @@ private void doScan(Set<String> classNames) {
...
+                        for (String inExt : builderParams.inExts()) {
                             extToBuilder.put(inExt, (Class<? extends Builder<?>>) klass);
                         }
+
+                        ProtoParams protoParams = klass.getAnnotation(ProtoParams.class);
+                        if (protoParams != null) {
+                            ProtoBuilder.addMessageClass(builderParams.outExt(), protoParams.messageClass());
+
+                            for (String ext : builderParams.inExts()) {
+                                Class<?> inputKlass = protoParams.srcClass();
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java:238]: Is it intentional that it is named inputKlass?

**Review by britzl — DISMISSED** [review]

**Pull request merged**